File: sdks/typescript/test/conformance.test.ts
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import test from 'node:test'

import {
  audioFrameToJson,
  parseAudioFrame,
  parseRuntimeEvent,
  parseSessionControl,
  parseVoiceSource,
  runtimeEventToJson,
  sessionControlToJson,
  voiceSourceToJson,
  type AudioFrame,
  type RuntimeEvent,
  type SessionControl,
  type VoiceSource
} from '../src/index.ts'

type Message = 'RuntimeEvent' | 'SessionControl' | 'AudioFrame' | 'VoiceSource'
type Parsed = RuntimeEvent | SessionControl | AudioFrame | VoiceSource
interface FixtureSet {
  name: string
  message: Message
  path: string
  lines: number
  expect: 'accept' | 'reject'
  reason?: string
  rejection?: 'parse' | 'order'
  rejectLine?: number
  unknownFields?: string[]
}

const conformanceUrl = new URL('../../../conformance/', import.meta.url)
const manifest = JSON.parse(readFileSync(new URL('manifest.json', conformanceUrl), 'utf8')) as {
  fixtureSets: FixtureSet[]
}

function parse(message: Message, input: unknown): Parsed {
  switch (message) {
    case 'RuntimeEvent': return parseRuntimeEvent(input)
    case 'SessionControl': return parseSessionControl(input)
    case 'AudioFrame': return parseAudioFrame(input)
    case 'VoiceSource': return parseVoiceSource(input)
  }
}

function serialize(message: Message, value: Parsed): Record<string, unknown> {
  switch (message) {
    case 'RuntimeEvent': return runtimeEventToJson(value as RuntimeEvent)
    case 'SessionControl': return sessionControlToJson(value as SessionControl)
    case 'AudioFrame': return audioFrameToJson(value as AudioFrame)
    case 'VoiceSource': return voiceSourceToJson(value as VoiceSource)
  }
}

function orderingKey(message: Message, value: Parsed): bigint | undefined {
  switch (message) {
    case 'RuntimeEvent': return (value as RuntimeEvent).sequence
    case 'SessionControl': return (value as SessionControl).requestSequence
    case 'AudioFrame': return (value as AudioFrame).sequence
    case 'VoiceSource': return undefined
  }
}

function removePath(value: Record<string, unknown>, path: string): void {
  const parts = path.split('/')
  let current: unknown = value
  for (const part of parts.slice(0, -1)) {
    if (typeof current !== 'object' || current === null || Array.isArray(current)) return
    current = (current as Record<string, unknown>)[part]
  }
  if (typeof current === 'object' && current !== null && !Array.isArray(current)) {
    delete (current as Record<string, unknown>)[parts.at(-1) ?? '']
  }
}

test('all manifest conformance sets', () => {
  for (const fixtureSet of manifest.fixtureSets) {
    const lines = readFileSync(new URL(fixtureSet.path, conformanceUrl), 'utf8')
      .split('\n')
      .filter((line) => line.trim() !== '')
    assert.equal(lines.length, fixtureSet.lines, `typescript · ${fixtureSet.name} · count`)
    let previous: bigint | undefined
    const rejectLine = fixtureSet.rejectLine ?? 1
    lines.forEach((line, index) => {
      const lineNumber = index + 1
      const label = `typescript · ${fixtureSet.name} · line ${lineNumber}`
      const input = JSON.parse(line) as Record<string, unknown>
      let parsed: Parsed
      try {
        parsed = parse(fixtureSet.message, input)
      } catch (error) {
        const shouldReject = fixtureSet.expect === 'reject' && fixtureSet.rejection === 'parse' && lineNumber >= rejectLine
        assert.ok(shouldReject, `${label} · ${String(error)}`)
        return
      }
      if (fixtureSet.expect === 'reject' && fixtureSet.rejection === 'parse' && lineNumber >= rejectLine) {
        assert.fail(`${label} · expected ${fixtureSet.reason ?? 'rejection'}`)
      }

      const current = orderingKey(fixtureSet.message, parsed)
      if (current !== undefined) {
        const ordered = previous === undefined || current > previous
        const rejectsOrder = fixtureSet.expect === 'reject' && fixtureSet.rejection === 'order' && lineNumber >= rejectLine
        assert.equal(ordered, !rejectsOrder, `${label} · ${fixtureSet.reason ?? 'ordering'}`)
        if (ordered) previous = current
      }

      const expected = structuredClone(input)
      const actual = structuredClone(serialize(fixtureSet.message, parsed))
      for (const path of fixtureSet.unknownFields ?? []) {
        removePath(expected, path)
        removePath(actual, path)
      }
      assert.deepEqual(actual, expected, `${label} · round-trip`)
    })
  }
})

File: sdks/dart/murmur_protocol/test/conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:murmur_protocol/murmur_protocol.dart';
import 'package:test/test.dart';

void main() {
  test('passes every manifest conformance set', () {
    final conformance = Directory('../../../conformance');
    final manifest = requireObject(
      jsonDecode(File('${conformance.path}/manifest.json').readAsStringSync()),
      'manifest',
    );
    final fixtureSets = manifest['fixtureSets']! as List<Object?>;
    for (final fixtureValue in fixtureSets) {
      final fixtureSet = requireObject(fixtureValue, 'fixture set');
      final name = fixtureSet['name']! as String;
      final message = fixtureSet['message']! as String;
      final lines = File(
        '${conformance.path}/${fixtureSet['path']}',
      ).readAsLinesSync().where((line) => line.trim().isNotEmpty).toList();
      expect(
        lines,
        hasLength(fixtureSet['lines']! as int),
        reason: 'dart · $name · count',
      );
      BigInt? previous;
      final rejectLine = fixtureSet['rejectLine'] as int? ?? 1;
      for (var index = 0; index < lines.length; index++) {
        final lineNumber = index + 1;
        final label = 'dart · $name · line $lineNumber';
        final input = requireObject(jsonDecode(lines[index]), message);
        Object parsed;
        try {
          parsed = _parse(message, input);
        } on Object catch (error) {
          final shouldReject =
              fixtureSet['expect'] == 'reject' &&
              fixtureSet['rejection'] == 'parse' &&
              lineNumber >= rejectLine;
          expect(shouldReject, isTrue, reason: '$label · $error');
          continue;
        }
        if (fixtureSet['expect'] == 'reject' &&
            fixtureSet['rejection'] == 'parse' &&
            lineNumber >= rejectLine) {
          fail('$label · expected ${fixtureSet['reason']}');
        }
        final current = _orderingKey(parsed);
        if (current != null) {
          final ordered = previous == null || current > previous;
          final rejectsOrder =
              fixtureSet['expect'] == 'reject' &&
              fixtureSet['rejection'] == 'order' &&
              lineNumber >= rejectLine;
          expect(
            ordered,
            !rejectsOrder,
            reason: '$label · ${fixtureSet['reason'] ?? 'ordering'}',
          );
          if (ordered) previous = current;
        }
        final expected = requireObject(jsonDecode(lines[index]), message);
        final actual = requireObject(
          jsonDecode(jsonEncode(_serialize(parsed))),
          message,
        );
        final unknownFields =
            fixtureSet['unknownFields'] as List<Object?>? ?? const [];
        for (final path in unknownFields.cast<String>()) {
          _removePath(expected, path);
          _removePath(actual, path);
        }
        expect(actual, equals(expected), reason: '$label · round-trip');
      }
    }
  });

  test('models a source without Flutter or a connector SDK', () {
    final source = VoiceSource(
      id: 'synthetic-1',
      displayName: 'Synthetic microphone',
      transport: VoiceSourceTransport.synthetic,
      capabilities: {VoiceSourceCapability.liveAudio},
    );
    expect(source.supports(VoiceSourceCapability.liveAudio), isTrue);
    expect(source.supports(VoiceSourceCapability.battery), isFalse);
  });
}

Object _parse(String message, Map<String, Object?> input) => switch (message) {
  'RuntimeEvent' => RuntimeEvent.fromJson(input),
  'SessionControl' => SessionControl.fromJson(input),
  'AudioFrame' => AudioFrame.fromJson(input),
  'VoiceSource' => VoiceSource.fromJson(input),
  _ => throw FormatException('unknown message $message'),
};

Map<String, Object?> _serialize(Object value) => switch (value) {
  final RuntimeEvent event => event.toJson(),
  final SessionControl control => control.toJson(),
  final AudioFrame frame => frame.toJson(),
  final VoiceSource source => source.toJson(),
  _ => throw FormatException('unknown parsed type ${value.runtimeType}'),
};

BigInt? _orderingKey(Object value) => switch (value) {
  final RuntimeEvent event => event.sequence,
  final SessionControl control => control.requestSequence,
  final AudioFrame frame => frame.sequence,
  _ => null,
};

void _removePath(Map<String, Object?> value, String path) {
  final parts = path.split('/');
  Map<String, Object?>? current = value;
  for (final part in parts.take(parts.length - 1)) {
    final next = current?[part];
    if (next is! Map<String, Object?>) return;
    current = next;
  }
  current?.remove(parts.last);
}
